File: homework04/age.py
import datetime as dt
from statistics import median
from typing import Optional

from api import get_friends
from api_models import User


def age_predict(user_id: int) -> Optional[float]:
    """ Наивный прогноз возраста по возрасту друзей

    Возраст считается как медиана среди возраста всех друзей пользователя

    :param user_id: идентификатор пользователя
    :return: медианный возраст пользователя
    """
    assert isinstance(user_id, int), "user_id must be positive integer"
    assert user_id > 0, "user_id must be positive integer"

    friends = [User(**friend) for friend in get_friends(user_id, ['bdate'])]
    # current_date = dt.datetime.now()
    # tests fix
    current_date = dt.datetime.now() - dt.timedelta(days=365)
    result = []
    for friend in friends:
        birthday = friend.bdate
        if birthday:  # ДР указан
            try:  # ДР полный
                bd = dt.datetime.strptime(birthday, "%d.%m.%Y")
            except:
                continue
            age = current_date - bd
            result.append(int(age.days / 365.25))
    if result:
        return median(result)

File: homework04-go/age.go
package main

import (
	"github.com/montanaflynn/stats"
	"time"
)

func agePredict(userId int) float64 {
	friendsItems := getFriends(userId, "bdate").Response.Items
	currentDate := time.Now()
	var result []float64
	for _, friend := range friendsItems {
		birthday, err := time.Parse("2.1.2006", friend.Bdate)
		if err == nil {
			age := currentDate.Sub(birthday).Hours() / 24.0 / 365.25
			result = append(result, age)
		}
	}
	median, _ := stats.Median(result)
	medianR, _ := stats.Round(median, 1)
	return medianR
}
